packages/database/src/core/util/ServerValues.ts

```diff
@@ -26,6 +26,53 @@ import { ChildrenNode } from '../snap/ChildrenNode';
 import { SyncTree } from '../SyncTree';
 import { Indexable } from './misc';
 
+/* It's critical for performance that we do not calculate actual values from a SyncTree
+ * unless and until the value is needed. Because we expose both a SyncTree and Node
+ * version of deferred value resolution, we ned a wrapper class that will let us share
+ * code.
+ *
+ * @see https://github.com/firebase/firebase-js-sdk/issues/2487
+ */
+interface DeferredExistingValue {
+  getImmediateChild(childName: string): DeferredExistingValue;
+  node(): Node;
+}
+
+class ExistingSnapshotValue implements DeferredExistingValue {
+  private node_: Node;
+  constructor(node: Node) {
```

Comment:
Nit: You can drop the suffix for members of classes that are not part of the public API, which will allow you to use constructor property assignments (`constructor(readaonly node:Node)`).

Renaming to `node` also allows you to get rid of the getter here.

Reply:
If I do that I fail to implement the protocol because `node` is a value instead of a method.